core/grafana-data-source/src/util.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+/// Get the current unix timestamp in milliseconds.
+pub fn now_millis() -> i64 {
+	chrono::Utc::now().timestamp_millis()
+}
+
+// find the index of a timestamp
+pub fn find_index(slice: &[(f32, i64)], timestamp: i64) -> usize {
+	slice.binary_search_by_key(&timestamp, |&(_, timestamp)| timestamp)
+		.unwrap_or_else(|index| index)
+}
+
+// Evenly select `num_points` points from a slice
+pub fn select_points<T: Copy>(slice: &[T], num_points: usize) -> Vec<T> {
+	(0 .. num_points - 1)
+		.map(|i| slice[i * slice.len() / (num_points - 1)])
```

Comment:
The server will panic if the user does a query with `num_points` equal to 0 or 1. We should try to avoid any "raw" arithmetic.